### PR TITLE
[PUBDEV-8809] Add Retry Logic to R Client to Prevent Temporary Failures

### PR DESCRIPTION
**Pull request by mn-mikke**

### CODE REVIEW

File: h2o-r/h2o-package/R/communication.R [hunk 1]
```diff
@@ -507,58 +507,75 @@
...
+    if (is.null(stopError)) {
+      return(rv$payload)
+    } else if (number_of_tries <= 0) {
+      stop(stopError)
     } else {
-      stop("Unexpected HTTP Status code")
-    }
+      cat("\n")
+      cat(sprintf("Error occurred, retrying in %d seconds", retry_interval))
+      cat("\n")
```
Comment (tomasfryda) [inline h2o-r/h2o-package/R/communication.R:575]: I'd probably use `warning` (or `message`) so it is possible to suppress it / redirect to a file etc.

File: h2o-r/h2o-package/R/communication.R [hunk 2]
```diff
@@ -1045,21 +1062,39 @@ h2o.get_job <- function(job_key, jobPollSuccess = FALSE, jobIsRecoverable = FALS
...
+    }
+
+    if (is.null(stopError)) {
+      break
+    } else if (number_of_tries <= 0) {
+      stop(stopError)
+    } else {
+      cat("\n")
+      cat(sprintf("Error occurred, retrying in %d seconds", retry_interval))
+      cat("\n")
```
Comment (tomasfryda) [inline h2o-r/h2o-package/R/communication.R:1095]: Use `warning` here instead of `cat`.

File: h2o-r/h2o-package/R/communication.R [hunk 3]
```diff
@@ -507,58 +507,75 @@
...
-      cat(sprintf("%s\n", exceptionType))
-    }
-
-    stacktrace = jsonObject$stacktrace
-    if (! is.null(stacktrace)) {
-      print(jsonObject$stacktrace)
+      stopError <- sprintf("Unexpected CURL error: %s", errorMessage)
+    } else if (rv$httpStatusCode != 200) {
+      cat("\n")
+      cat(sprintf("ERROR: Unexpected HTTP Status code: %d %s (url = %s)\n", rv$httpStatusCode, rv$httpStatusMessage, rv$url))
```
Comment (tomasfryda) [inline h2o-r/h2o-package/R/communication.R:538]: Use `warning` here instead of `cat`.

**Comment by tomasfryda** [comment]
I would prefer to use exponential backoff for the delay but since we didn't have any retry logic it would probably not make much of a difference. But it might be worth considering.

File: h2o-r/h2o-package/R/communication.R (hunk 3 above)
Comment (mn-mikke) [reply h2o-r/h2o-package/R/communication.R]: will change, thanks!

File: h2o-r/h2o-package/R/communication.R (hunk 1 above)
Comment (mn-mikke) [reply h2o-r/h2o-package/R/communication.R]: will change, thanks

File: h2o-r/h2o-package/R/communication.R (hunk 2 above)
Comment (mn-mikke) [reply h2o-r/h2o-package/R/communication.R]: will change, thanks!

**Comment by mn-mikke** [comment]
> I would prefer to use exponential backoff for the delay but since we didn't have any retry logic it would probably not make much of a difference. But it might be worth considering.

@tomasfryda I think it's good idea if the retry logic was enabled by default. I propose to have the retry logic disabled by default in order to let the R client fail fast if anything goes wrong. If users have unstable connectivity from R client to h2o leader node, they could enable the retry logic and set the delay according to their needs. This just my proposal, I'm open to change  if there are more people with the same view. 

cc @krasinski @sebhrusen any opinion?

**Comment by sebhrusen** [comment]
> I think it's good idea if the retry logic was enabled by default. I propose to have the retry logic disabled by default in order to let the R client fail fast if anything goes wrong. If users have unstable connectivity from R client to h2o leader node, they could enable the retry logic and set the delay according to their needs. This just my proposal, I'm open to change if there are more people with the same view.

When implementing retry logic, you don't usually retry on any kind of error, I would also externalize the codes for which you want to retry (I guess you at least want 408 and maybe 504) and retry only on those by default, which would also allow you to enable the retry logic by default with more confidence: for timeouts, I'd also recommend exponential backoff by default.
Maybe you're also targetting some other specific errors with this PR, idk.

**Review by michalkurka — CHANGES_REQUESTED** [review]
This change seems to affect all types of requests - GETs, POSTs...

I think it is generally not possible to make a retry work for requests that change the state on the backend - eg. simple x = x + 1 cannot be replayed because you don't know if the backend is 

A) didn't get the request at all,
B) failed to execute the request and correctly send an error back,
C) it is still working on the request but the TCP/IP or HTTP timed out in the meantime,
D) successfully completed the request but the client failed to receive a response.

**Comment by michalkurka** [comment]
Closing for now - not needed anymore

**Pull request closed without merging**